sycl/source/detail/graph_impl.cpp

```diff
@@ -77,29 +77,195 @@ void duplicateNode(const std::shared_ptr<node_impl> Node,
+    std::list<std::shared_ptr<node_impl>> &HostTaskList) {
+  if (Node->MCGType == sycl::detail::CG::CGTYPE::CodeplayHostTask) {
+    if (Node->MPartitionNum != -1) {
+      HostTaskList.push_front(Node);
+    }
+    return;
+  }
+  Node->MPartitionNum = PartitionNum;
+  for (auto &Successor : Node->MSuccessors) {
+    propagatePartitionDown(Successor.lock(), PartitionNum, HostTaskList);
+  }
+}
+
+/// Tests if the node is a root of its partition (i.e. no predecessors that
+/// belong to the same partition)
+/// @param Node node to test
+/// @return True is `Node` is a root of its partition
+bool isPartitionRoot(std::shared_ptr<node_impl> Node) {
+  for (auto Predecessor : Node->MPredecessors) {
+    if (Predecessor.lock()->MPartitionNum == Node->MPartitionNum) {
+      return false;
+    }
+  }
+  return true;
+}
+
 } // anonymous namespace
 
-void exec_graph_impl::schedule() {
+void partition::schedule() {
   if (MSchedule.empty()) {
-    for (auto &Node : MGraphImpl->MRoots) {
-      sortTopological(Node.lock(), MSchedule);
+    for (auto &Node : MRoots) {
+      sortTopological(Node.lock(), MSchedule, true);
     }
   }
 }
 
+void exec_graph_impl::makePartitions() {
+  int CurrentPartition = -1;
+  std::list<std::shared_ptr<node_impl>> HostTaskList;
+  // find HostTask
+  for (auto Node : MGraphImpl->MNodeStorage) {
+    if (Node->MCGType == sycl::detail::CG::CodeplayHostTask) {
+      HostTaskList.push_back(Node);
+    }
+  }
+
+  // annotate nodes
+  // The first step in graph partitioning is to annotate all nodes of the graph
+  // with a temporary partition or group number. This step allows us to group
+  // the graph nodes into sets of nodes with kind of meta-dependencies that must
+  // be enforced by the runtime. For example, Group 2 depends on Groups 0 and 1,
+  // which means that we should not try to run Group 2 before Groups 0 and 1
+  // have finished executing. Since host-tasks are currently the only tasks that
+  // require runtime dependency handling, groups of nodes are created from
+  // host-task nodes. We therefore loop over all the host-task nodes, and for
+  // each node :
+  //  - Its predecessors are assigned to group number `n-1`
+  //  - The node itself constitutes a group, group number `n`
+  //  - Its successors are assigned to group number `n+1`
+  // Since runnning multiple partitions slows down the whole graph execution, we
+  // then try to reduce the number of partition by merging them when possible.
+  // Typically, in some cases, the grouping algorithm can create two partitions
+  // of target nodes if a host-task `A`, which is stored before in the nodes
+  // list, is a successor of another host-task `B` (this case can only happen if
+  // users use the make_edge function.). In this case, the host-task node `A` is
+  // reprocessed after the node `B` and group that inclued the predecessor of
+  // `B` can be merged with the group of the predecessors of the node `A`.
```

Comment:
```suggestion
  // users use the make_edge function.). In this case, the host-task node `A` is
  // reprocessed after node `B` and the group that includes the predecessor of
  // `B` can be merged with the group of the predecessors of node `A`.
```
Maybe we can simplify this sentences a little bit to make it easier to follow. Breaking up into 2 sentences?

Reply:
I rewrote this part. I hope it is clearer now.